### DISCUSSION TITLE
Is it possible to map C# `object` type in Slice?

### DISCUSSION THREAD

**Discussion by babaq** [zeroc-ice/ice#3214] [Q&A] [answered]
I plan to expose a backend written in C# using ICE. Part of the planed API would be:

```CSharp
bool SetParam(string name, object value)
object GetParam(string name)
```

How to map the C# API to Slice?

A param can be variety of types. In protobuf, there is a `Any` type that fit my requirement, but i can't find a way to implement it in Slice.

**Comment by bernardnormier** [comment]
If your focus is C#, you may want to check out [IceRPC](https://zeroc.com/icerpc).

In Ice 3.7 and prior releases, we support "object serialization" in C#, which allows you to use C# serialization to marshal any C# object into a sequence of bytes. See:
https://doc.zeroc.com/ice/3.7/the-slice-language/serializable-objects
https://doc.zeroc.com/ice/3.7/language-mappings/c-sharp-mapping/client-side-slice-to-c-sharp-mapping/serializable-objects-in-c-sharp

However, I would recommend against using this feature because:
 - object serialization is deprecated in .NET (even removed in the latest .NET versions)
 - we're dropping this feature in Ice 3.8.

In Ice - Slice, you can model "any" Slice type using the Value keyword - the base class for all Slice classes. You could convert your C# API into:
```slice
bool SetParam(string name, Value v);
Value GetParam(string name);
```

And then define a number of Slice classes that hold the various types you want to get/set with Get/SetParam:
```slice
class StringValue { string v; } // holds a string 
class ColorValue { Color v; }  // holds a Color enumerator
...
```
 
[New Slice](https://docs.icerpc.dev/slice2) provides a more elegant and efficient solution: [enum with fields](https://docs.icerpc.dev/slice2/language-guide/enum-types#enum-with-fields)

```slice
enum Shape {
    Circle(radius: uint32)
    Rectangle(width: uint32, length: uint32)
    Dot
}
```

It's nicer than Protobuf's Any type, and it could be better suited for your use-case.

**Reply by babaq** [reply]
@bernardnormier Thanks for the quick response. 

Our current situation would be a C# Server(.NET Framework 4.8) and Python/C# Clients. So if i use IceRPC at server in Slice1 mode, Python client would be able to interop, right?

**Reply by bernardnormier** [reply, marked as answer]
Yes, IceRPC with Slice1 and the ice protocol interoperates with Ice except for a few uncommon situations (you use compression with Ice, or you use batch requests with Ice). For most applications, there is no interop issue.

However, IceRPC C# requires a recent .NET (.NET 8.0 or 9.0). It won't work with .NET Framework 4.8.

**Reply by babaq** [reply]
We will try ICE instead, because we are currently stuck with .NET Framework. Thanks.
